dali/operators/reader/loader/filesystem.cc

```diff
@@ -47,168 +47,116 @@ std::string join_path(const std::string &dir, const std::string &path) {
+
+  DIR *dir = opendir(parent_dir.c_str());
+  if (dir == nullptr)
+    return false;
 
   dirent *entry;
 
-  if (filter.empty()) {
-    while ((entry = readdir(dir))) {
+  while ((entry = readdir(dir))) {
 #ifdef _DIRENT_HAVE_D_TYPE
     /*
      * we support only regular files and symlinks, if FS returns DT_UNKNOWN
      * it doesn't mean anything and let us validate filename itself
      */
-      if (entry->d_type != DT_REG && entry->d_type != DT_LNK &&
-         entry->d_type != DT_UNKNOWN) {
-        continue;
-      }
-#endif
-      std::string rel_path = curr_entry + dir_sep + std::string{entry->d_name};
-      if (HasKnownExtension(std::string(entry->d_name))) {
-         file_list.push_back(rel_path);
-      }
+    if (entry->d_type != DT_REG && entry->d_type != DT_LNK &&
+        entry->d_type != DT_UNKNOWN) {
+      continue;
     }
-  } else {
-    // use glob to do the file search
-    glob_t pglob;
-    std::string pattern = curr_dir_path + dir_sep + filter;
-    if (glob(pattern.c_str(), GLOB_TILDE, NULL, &pglob) == 0) {
-      // iterate through the matched files
-      for (unsigned int count = 0; count < pglob.gl_pathc; ++count) {
-        std::string match(pglob.gl_pathv[count]);
-        std::string rel_path = curr_entry + dir_sep + match.substr(match.find_last_of(dir_sep)+1);
-        file_list.push_back(rel_path);
+#endif
+    std::string fname(entry->d_name);
+    for (auto &filter : filters) {
+      if (fnmatch(filter.c_str(), fname.c_str(), case_sensitive_filter ? 0 : FNM_CASEFOLD) == 0) {
+        files.emplace_back(fname);
+        break;
       }
-      // clean up
-      globfree(&pglob);
     }
   }
   closedir(dir);
+  return true;
 }
 
-
-vector<std::string> traverse_directories(const std::string &file_root, const std::string &filter) {
+vector<std::pair<string, int>> traverse_directories(const std::string &file_root,
+                                                    const std::vector<std::string> &filters,
+                                                    const bool case_sensitive_filter) {
   // open the root
-  DIR *dir = opendir(file_root.c_str());
+  std::vector<std::string> subdirs;
+  bool ret = list_subdirectories(subdirs, file_root);
+  DALI_ENFORCE(ret, "Directory " + file_root + " could not be opened.");
 
-  DALI_ENFORCE(dir != nullptr,
-      "Directory " + file_root + " could not be opened.");
-
-  struct dirent *entry;
+  std::vector<std::pair<std::string, int>> file_label_pairs;
+  std::vector<std::string> files;
+  for (unsigned dir_count = 0; dir_count < subdirs.size(); ++dir_count) {
+    files.clear();
+    auto dirpath = join_path(file_root, subdirs[dir_count]);
+    list_files(files, dirpath, filters, case_sensitive_filter);
+    for (const auto& f : files) {
+      file_label_pairs.push_back({join_path(dirpath, f), dir_count});
+    }
+  }
+  // sort file names as well
+  std::sort(file_label_pairs.begin(), file_label_pairs.end());
```

Comment:
Shouldn't that be a job of `list_files`? If the files returned by `list_files` were sorted then with the subdirs already sorted you'd have a guarantee of the paths being sorted - and it would in fact be cheaper.
For D directories and F files you'd have:
```O(D log D + DF log F)```

Now you have:
```
O(DF log DF) == O(DF log D + DF log F))
                   ^____ unnecessary re-sorting of directory names
```